Fixes in 'tosa.reshape' lowering and folder

rafaelubalmw opened the pull request:
This pull request addresses missing features in the lowering conversion pattern for the `tosa.reshape` op and a bug in its canonicalizer.

- Example of a valid use of `tosa.reshape` previously not supported:

```
func.func @main(%input: tensor<?x?x?xf32>) -> tensor<1x3x2x1xf32> {
  %0 = tosa.reshape %input {new_shape = array<i64: 1, 3, 2, 1>} : (tensor<?x?x?xf32>) -> tensor<1x3x2x1xf32>
  return %0 : tensor<1x3x2x1xf32>
}
```

- The new lowering is based on the use of `tensor.reshape` instead of a combination of `tensor.collapse_shape` + `tensor.expand_shape`.

- When no -1 placeholder is present in the `new_shape` attribute, the target shape is encoded with an `arith.constant` op and the reshape occurs with a `tensor.reshape` op.

- When a -1 placeholder is used in `new_shape` and the corresponding dimension in the result type is dynamic, the missing dimension size is inferred by calculating the input tensor size (`tensor.collapse_shape` + `tensor.dim`) and dividing it by the product of all other target dimension sizes (`arith.divui`).

- When a -1 placeholder is used in `new_shape` and the corresponding dimension in the result type is static, the missing dimension size is grabbed from the result type.

- Fixed bug in canonicalization pattern `tosa::ReshapeOp::fold()`. The input and result types being equal is not a sufficient condition for folding. If there is more than 1 dynamic dimension in the input and result types, a productive reshape could still occur. Unit tests are now available for 1 (existing test) and 2 (new test) dynamic dimensions.

rafaelubalmw committed:
Lowering of 'tosa.reshape' based on 'tensor.reshape'

rafaelubalmw committed:
Bux fixed in 'tosa.reshape' canonicalization. Completed lowering for 'tosa.reshape' lowering

krzysz00 commented on mlir/lib/Conversion/TosaToTensor/TosaToTensor.cpp:
```diff
@@ -19,217 +19,98 @@
+static Value getTensorSize(OpBuilder& builder, Location loc, TypedValue<TensorType> input) {
+  // If the input tensor is statically shaped, return its size as a constant.
+  if (input.getType().hasStaticShape()) {
+    auto shape = input.getType().getShape();
```
This is `getNumElements()`

sabauma commented:
I think it is worth considering the downstream implications of changing the lowering from `tensor.collapse_shape`/`tensor.expand_shape` to `tensor.reshape`. Some existing optimizations rely on the fact that `tensor.expand_shape` and `tensor.collapse_shape` are easier to reason about. The `--linalg-fuse-elementwise-ops` pass and its associated rewrite patterns comes to mind as a good example.

For instance, the following example is able to produce a single fused `linalg.generic` with the existing lowering, but will no longer fuse with this change.

```mlir
// RUN: mlir-opt %s --pass-pipeline="builtin.module(func.func(tosa-to-tensor,tosa-to-linalg,linalg-fuse-elementwise-ops))"

func.func @main(%arg0: tensor<100xf32>, %arg1: tensor<100xf32>) -> tensor<10x10xf32> {
  %add1 = tosa.add %arg0, %arg1 : (tensor<100xf32>, tensor<100xf32>) -> (tensor<100xf32>)

  %2 = "tosa.reshape"(%add1) {new_shape = array<i64: 10, 10>} : (tensor<100xf32>) -> tensor<10x10xf32>
  %3 = "tosa.reshape"(%arg1) {new_shape = array<i64: 10, 10>} : (tensor<100xf32>) -> tensor<10x10xf32>

  %add2 = tosa.add %2, %3 : (tensor<10x10xf32>, tensor<10x10xf32>) -> (tensor<10x10xf32>)
  return %add2 : tensor<10x10xf32>
}
```

There may be other fallout that other reviewers are aware of. Maybe @eric-k256 or @rsuderman could chime in with more examples, or know those who would know.

krzysz00 commented:
I'll note that I am one of those downstreams who's relying on collapse/expand pairs for `tosa.reshape`, but I didn't raise it because I figured it'd be resolvable.

If there isn't already a pass that rewrites `tensor.reshape` to `tensor.collapse_shape` and `tensor.expand_shape` where possible, we might want to add one, since that could then be slid into the Tosa to Linalg pipeline to preserve the existing behavior.

rafaelubalmw committed:
Addressed feedback, reverting to collapse-expand pattern

rafaelubalmw commented:
@sabauma @krzysz00: Thank you very much for your insightful feedback. The advantages of sticking to the collapse-expand pattern seemed a good enough reason to revert my lowering strategy and instead focus on enhancing it to support all `tosa.reshape` cases. The new version differs almost completely from my original PR, so I'd appreciate if you could take another look. Here's a brief description:

- I modified existing auxiliary functions to avoid double-checking for invariants guaranteed by the `tosa.reshape` op verifier, such as target shape compatibility. At conversion, these invariants are checked with assert's rather than pattern match failures.

- Previously unsupported cases were caused by failures to guarantee type consistency in the emitted `tensor.expand_shape` op (i.e., input and output shapes must be both static or both dynamic). The result type for this op is now manually constructed, and an additional `tensor.cast` op is emitted if this type differs from the `tosa.reshape` result type.

- An extended set of tests are intended to cover relevant conversion paths. Tests are named using patten `test_reshape_<rank>_{up|down|same}_{s2s|s2d|d2s|d2d}_{explicit|auto}[_empty][_identity]`, where:

  - `<rank>` is the input rank (e.g., 3d, 6d)
  - `{up|down|same}` indicates whether the reshape increases, decreases, or retains the input rank.
  - `{s2s|s2d|d2s|d2d}` indicates whether reshape converts a statically shaped input to a statically shaped result (`s2s`), a statically shaped input to a dynamically shaped result (`s2d`), etc.
  - `{explicit|auto}` is used to indicate that all values in the `new_shape` attribute are >=0 (`explicit`) or that a -1 placeholder value is used (`auto`).
  - `empty` is used to indicate that `new_shape` includes a component set to 0.
  - `identity` is used when the input and result shapes are the same.

Other considerations regarding @krzysz00, to the extent that it is still applicable:

- I created tensor types using `clone()`, as suggested.

- I avoided the use of `auto` for basic types. I agree it helps to spell these out, especially when it comes to integer sizes. I did leave all other uses of `auto` (`SmallVector`, `Location`, `Value`, ...) as it seems consistent with other occurrences in the repo.

sabauma commented on mlir/lib/Conversion/TosaToTensor/TosaToTensor.cpp:
```diff
@@ -19,24 +19,77 @@
+
+    // Calculate the product of all elements in 'newShape' except for the -1
+    // placeholder, which we discard by negating the result.
+    int64_t totalSizeNoPlaceholder = -std::accumulate(
+        newShape.begin(), newShape.end(), 1, std::multiplies());
```
Is it true that the assumption here that `newShape` contains at most a single -1? Could you add an assertion to that effect at the top of the function?

sjarus replied:
Thanks for this PR, @rafaelubalmw ! You're right to note the elimination of the explicit requirement that at most one dim be -1 . This was removed from the spec sometime ago (cc: @eric-k256) . 

Going forward, we plan to relax the shape parameter significantly. Any number of undetermined dims will be permitted, but by requiring the legalization to onboard dynamic shape resolution, potentially at runtime. 

The compile time resolvable construct will only permit up to one unknown dim, because as you said, that's all that makes sense.

sabauma commented on mlir/lib/Conversion/TosaToTensor/TosaToTensor.cpp:
```diff
@@ -19,24 +19,77 @@
+  // shaped input from being reshaped into a statically shaped result. We may
+  // simply turn the first result dimension dynamic to address this.
+  if (!inputIsStatic && resultIsStatic)
+    resultShape[0] = ShapedType::kDynamic;
```
Can `resultShape` be a rank-0 tensor?

rafaelubalmw replied:
It indeed could. But no more. See my latest comment below.

sabauma commented on mlir/lib/Conversion/TosaToTensor/TosaToTensor.cpp:
```diff
@@ -62,174 +115,113 @@ static bool findIntermediateShape(ArrayRef<int64_t> lhsShape,
-      exprs.push_back(rewriter.getAffineDimExpr(i));
-    reassociationMap = {exprs};
-    return true;
+    for (int64_t i = 0, s = srcShape.size(); i < s; ++i)
```
I generally find `llvm::seq` more readable than fiddly low-level loops.

sabauma commented on mlir/test/Conversion/TosaToTensor/tosa-to-tensor.mlir:
```diff
@@ -1,95 +1,282 @@
+// CHECK-SAME: %[[ARG_0:[a-zA-Z0-9_]+]]: tensor<f32>
+// CHECK: return %[[ARG_0]] : tensor<f32>
+func.func @test_reshape_0d_same_s2s_explicit(%arg0: tensor<f32>) -> tensor<f32> {
+  %0 = "tosa.reshape"(%arg0) {new_shape = array<i64>} : (tensor<f32>) -> tensor<f32>
```
Could you add a test case for a reshape of a multi-dimensional (size 1) tensor down do a rank-0 tensor. I don't see any case like that. Possibly an example of the reverse as well. In my experience, rank-0 tensors tend to be edge cases in a lot of conversions.

rafaelubalmw committed:
Added restriction for max number of missing target dims in op verifier

joker-eph commented on mlir/lib/Dialect/Tosa/IR/TosaOps.cpp:
```diff
@@ -970,6 +970,11 @@ mlir::LogicalResult tosa::ReshapeOp::verify() {
     }
   }
+
+  int missingDims = std::count(getNewShape().begin(), getNewShape().end(), -1);
```
```suggestion
  int missingDims = llvm::count(getNewShape(), -1);
```

rafaelubalmw committed:
Added support for 0D tensor reshapes and addressed feedback

rafaelubalmw commented:
As observed by @sabauma, 0D tensors were not properly handled. I updated the PR as follows:

- I added additional checks for this special case.

- I added the following unit tests:

  - `reshape_0d_up_s2s_explicit`
  - `reshape_0d_up_s2s_auto`
  - `reshape_0d_up_s2d_explicit`
  - `reshape_0d_up_s2d_auto`
  - `reshape_1d_down_s2s_explicit`
  - `reshape_1d_down_d2s_explicit`
  - `reshape_4d_down_d2s_explicit`

- The new code led me to the observation that bufferization does not properly handle a `tensor.collapse_shape` op producing a 0D tensor from a dynamically shaped one. It looks like `memref.collapse_shape` does not allow it. While the proper way to address this would involve releasing the `memref.collapse_shape` restriction and verifying correct bufferization for this case, I'll leave that as possible future work. For now, I just avoid this scenario by casting the input if necessary (see `inferReshapeInputType()`).

- I renamed all the type inference functions and slightly modified their interfaces for symmetry.

Who knew there was so much to consider for a simple tensor reshape... :) Please let me know what you think.

sabauma reviewed (approved):
Thanks for the followup. Looks good to me.

sjarus commented on mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp:
```diff
@@ -795,7 +795,10 @@ OpFoldResult ReshapeOp::fold(FoldAdaptor adaptor) {
+  // Fold when the input and output types are the same. This is only safe when
+  // there is at most 1 dynamic dimension. For 2 or more dynamic dimensions,
+  // there may still be a productive reshape.
+  if (inputTy == outputTy && inputTy.getNumDynamicDims() < 2)
```
As mentioned in the earlier comment, this restriction will be relaxed. This would probably necessitate dynamic cases to be predicated on something like tensor.dim for later resolution. 

However, for the purposes of this PR this code looks fine.

rafaelubalmw committed:
Merge branch 'main' into tosa-reshape-lowering

bjacob commented:
@rafaelubalmw @sabauma @sjarus , I have run into a regression https://github.com/llvm/llvm-project/issues/87396 while integrating in a downstream (IREE).
